scalafix-core/shared/src/main/scala/scalafix/internal/rule/SingleAbstractMethod.scala

```diff
@@ -0,0 +1,174 @@
+
+    // extract the top name position java.util.TimerTask => TimerTask
+    def parentNamePosition(tpe: Type): Option[Position] = {
+      tpe match {
+        case Type.Select(_, name) => Some(name.pos)
+        case name: Type.Name => Some(name.pos)
+        case _ => None
+      }
+    }
+
+    object Sam {
+      def unapply(tree: Tree): Option[(Term.Function, Type)] = {
+        tree match {
+          case Term.NewAnonymous(
+              Template(
+                _,
+                List(Init(clazz, _, _)),
+                _,
+                List(Defn.Def(_, method, _, List(params), _, body))
+              )
+              ) =>
+            val singleAbstractOverride =
+              (for {
+                definition <- index.denotation(method)
+                overrideSymbol <- definition.overrides.headOption
+                if definition.overrides.size == 1
+                overrideDefinition <- index.denotation(overrideSymbol)
+              } yield overrideDefinition.isAbstract).getOrElse(false)
+
+            val singleMember =
+              (for {
+                position <- parentNamePosition(clazz)
+                symbol <- index.symbol(position)
+                denfinition <- index.denotation(symbol)
+              } yield denfinition.members.size == 1).getOrElse(false)
+
+            if (singleAbstractOverride && singleMember) {
+              Some(
+                (
+                  Term.Function(params.map(_.copy(decltpe = None)), body),
+                  clazz
+                )
+              )
+            } else {
+              None
+            }
+
+          case _ => None
+        }
+      }
+    }
+
+    collectOnce(ctx.tree) {
+      case term @ Defn.Val(mods, List(Pat.Var(name)), tpe0, Sam(lambda, tpe)) =>
+        ctx.replaceTree(
+          term,
+          Defn
+            .Val(mods, List(Pat.Var(name)), Some(tpe0.getOrElse(tpe)), lambda)
+            .show[Syntax]
+        )
+      case term @ Defn.Var(
+            mods,
+            List(Pat.Var(name)),
+            tpe0,
+            Some(Sam(lambda, tpe))) =>
+        ctx.replaceTree(
+          term,
+          Defn
+            .Var(
+              mods,
+              List(Pat.Var(name)),
+              Some(tpe0.getOrElse(tpe)),
+              Some(lambda))
+            .show[Syntax]
+        )
+      case term @ Defn.Def(
+            mods,
+            name,
+            tparams,
+            paramss,
+            decltpe,
+            Sam(lambda, tpe)) =>
+        ctx.replaceTree(
+          term,
+          Defn
+            .Def(
+              mods,
+              name,
+              tparams,
+              paramss,
+              Some(decltpe.getOrElse(tpe)),
+              lambda)
+            .show[Syntax]
+        )
+
+      case tree @ Sam(lambda, _) =>
+        ctx.replaceTree(
+          tree,
+          lambda.show[Syntax]
+        )
+
+    }.asPatch
+  }
+
+  private def collectOnce[T](tree: Tree)(
```

Comment:
Why is this necessary? collect also uses `.lift`

Reply:
for example:

```scala
import java.lang.Runnable
class SingleAbstractMethod {
  val runnable1 = new Runnable(){ def run(): Unit = println("Run!")}
  new Thread(new Runnable(){ def run(): Unit = println("Hello, Thread!")})
}
```

```scala
ctx.tree.collect {
      case term @ Defn.Val(mods, List(Pat.Var(name)), tpe0, Sam(lambda, tpe)) => ...
      case tree @ Sam(lambda, _) => ... // [B]
}
```

`val runnable1` will be collected twice. first matching the Defn.Val then second matching the Sam.

Reply:
However, here is a counterexample:

```scala
  val runnable4 = new Runnable(){
    def run(): Unit = {
      val runnable5 = new Runnable(){ 
        def run(): Unit = println("runnable5!")
      }
    }
  }
```

Reply:
@olafurpg can you point to an example of this technique. I can’t envision it from your description alone but it seems like a technique I should know.

Reply:
@ShaneDelmore see https://github.com/scalacenter/scalafix/pull/594#discussion_r166322766 for an example. The trick is to do `super.apply` on the children that you know won't match against again against your pattern